### ISSUE TITLE
can't use bcrypt as the password check algorithm

### ISSUE THREAD

**Issue by oupala** [Kozea/Radicale#1111]
I am trying to make radicale works with an authentication htpasswd but with a bcrypt algorithm, with radicale is behind an apache using wsgi (on debian stable, in case you wonder which versions I'm using).

When I'm using a standard authentication algorithm (md5), everything works well. But when I switch to *bcrypt* algorithm, the authentication does not pass anymore.

Does anyone has an idea? This might be related to #901 and #785.

I'm asking myself if radical support all [bcrypt hash types](https://en.wikipedia.org/wiki/Bcrypt#Versioning_history)? Such as `2a`, `2b`, `2y`?

I'm not sure, but I feel like a hash with a `2y` prefix passes, but prefixes `2a` and `2b` do not pass... But I'm not an expert, so I might be completyl wrong in the way I'm trying to understand what's happening...

**Comment by pgnd** [comment]
this works here.

with
```
pip show Radicale
	Name: Radicale
	Version: 3.0.6
	Summary: CalDAV and CardDAV Server
	Home-page: https://radicale.org/
	Author: Guillaume Ayoub
	Author-email: guillaume.ayoub@kozea.fr
	License: GNU GPL v3
	Location: /usr/lib/python3.8/site-packages
	Requires: defusedxml, passlib, vobject, python-dateutil
	Required-by:
```

make sure you've got all your prereqs

for my fedora32 instance/needs, i've installed
```
python3-passlib
python3-bcrypt
python3-dateutil
python3-defusedxml
python3-vobject
python3-m2crypto
python3-pyOpenSSL
python3-ndg_httpsclient
python3-service-identity
```

ensure
```
	radicale.conf
		...
		[auth]
		...
+		htpasswd_filename = /path/to/password/file/passwords.sec
+		htpasswd_encryption = bcrypt
		...
```

create bcrypt passfile; e.g.,
```
_pass="123456789012345678901234567890123456789012345678"
_user="myusername"
_passfile="/path/to/password/file/passwords.sec"
_cost="14"
_hash=$( php -r "echo password_hash('$_pass', PASSWORD_BCRYPT, ['cost' => $_cost]) . PHP_EOL;" )
echo "$_user:$_hash" > $_passfile
```

many ways to that^ mountain ...

verify
```
htpasswd -vb $_passfile $_user "$_pass"
	Password for user myusername correct.
```

**Comment by oupala** [comment]
Thanks @pgnd for trying to help me with my problem.

I took some time, I took a step back, I rethinking about my problem and I realized that I was mixing multiple ways of using radicale.

The first thing that I omitted to mention is that I am using [radicale as WSGI](https://radicale.org/2.1.html#documentation/wsgi). And although radicale is used through apache2 (the web server), the radicale service was started.

The configuration of radicale was configured to [`htpasswd` as an authentication method](https://radicale.org/2.1.html#documentation/configuration/auth/type) while it should have been set to `remote_user` as the authentication is handled by apache and passed to radicale (*remote authentication* leads to `remote_user`).

As a consequence, as I am using radicale as a WSGI service, the authentication is handled by apache and I should have a look at apache2 documentation in order to find out if apache supports bcrypt algorithm for authentication.
